default.nix

```diff
@@ -1,116 +1,150 @@
+      rev = "dadb6a468b9cdc47442b48a47b848f8e8fbffda7";
+      sha256 = "YkUtM5Gos6xd7ZsZPqcuVy6DZqNA7n/exPfQngir+y0=";
     };
+    patches = [ ./stdlib-meta-update-imports.patch ];
+    meta = { };
+    libraryFile = "stdlib-meta.agda-lib";
+    everythingFile = "Main.agda";
+    buildInputs = [ agdaStdlib ];
   };
-  agda-ledger = agdaDeps.ledger;
-  agdaWithPkgs = agdaDeps.agda;
 
-  ledger-hs-src = stdenv.mkDerivation {
-    pname = "Agda-ledger-hs-src";
+  deps = [ agdaStdlib agdaFinset agdaStdlibMeta ];
+  agdaWithPkgs = customAgda.agda.withPackages { pkgs = deps; ghc = pkgs.ghc; };
+
+in
+rec {
+
+  agdaLedger = customAgda.agdaPackages.mkDerivation {
+    pname = "Agda-ledger";
     version = "0.1";
-    src = "${agda-ledger}";
-    meta = {};
-    buildInputs = [ agdaWithPkgs ];
-    buildPhase = "";
-    installPhase = ''
-      mkdir -p $out
-      agda -c --ghc-dont-call-ghc --compile-dir $out Ledger/Foreign/HSLedger.agda
-      cp Ledger/HSLedgerTest.hs $out
-      cp Ledger/agda-ledger-executable-spec.cabal $out
-      # Append all the modules generated by MAlonzo to the cabal file
-      find $out/MAlonzo -name "*.hs" -print | sed "s#^$out/#        #;s#\.hs##;s#/#.#g" >> $out/agda-ledger-executable-spec.cabal
-    '';
+    src = ./src;
+    meta = { };
+    buildInputs = deps;
+    postInstall = "cp -r latex $out";
+    extraExtensions = [ "hs" "cabal" ];
   };
-in {
-  agda = agdaWithPkgs;
-  agda-ledger = agda-ledger;
-  agda-ledger-executable-spec = haskellPackages.callCabal2nix "agda-ledger-executable-spec" "${ledger-hs-src}" {};
 
-  ledger-docs = stdenv.mkDerivation {
-    pname = "simple-utxo-ledger-docs";
-    version = "0.1";
-    src = "${agda-ledger}";
-    meta = {};
-    buildInputs = [ agdaWithPkgs (texlive.combine {
-      inherit (texlive)
-        scheme-small
-        xits
-        collection-latexextra
-        collection-latexrecommended
-        collection-mathscience
-        bclogo
-        latexmk;
-    })];
-    buildPhase = ''
-      ls Ledger/*.lagda | xargs -n 1 agda --latex
-      cd latex && latexmk -xelatex Ledger/Ledger.tex && cd ..
-    '';
-    installPhase = ''
-      mkdir -p $out
-      agda --html --html-dir $out/html Ledger/Ledger.lagda
-      cp latex/Ledger.pdf $out
-    '';
+  hsSrc = { dir, agdaLedgerPath, hsMainPath }:
+    stdenv.mkDerivation {
+      pname = "Agda-ledger-${dir}-hs-src";
+      version = "0.1";
+      src = "${agdaLedger}";
+      meta = { };
+      buildInputs = [ agdaWithPkgs ];
+      buildPhase = "";
+      installPhase = ''
+        mkdir -p $out
+        agda -c --ghc-dont-call-ghc --compile-dir $out ${agdaLedgerPath}
+        cp ${hsMainPath} $out
+        cp ${dir}/agda-ledger-executable-spec.cabal $out
+        # Append all the modules generated by MAlonzo to the cabal file
+        find $out/MAlonzo -name "*.hs" -print | sed "s#^$out/#        #;s#\.hs##;s#/#.#g" >> $out/agda-ledger-executable-spec.cabal
+      '';
+    };
+
+  ledgerHsSrc = hsSrc {
```

Comment:
There's a bunch of repetition here. What do you think about making a single function that accepts this set and generates all three derivations at once?

Reply:
Indeed! Let's see how that looks like, trying it now